rasa/engine/training/graph_trainer.py

```diff
@@ -0,0 +1,226 @@
+import copy
+import logging
+from pathlib import Path
+from typing import Any, Dict, Text, Type, Union
+
+from rasa.engine.caching import TrainingCache
+from rasa.engine.graph import ExecutionContext, GraphSchema
+from rasa.engine.runner.interface import GraphRunner
+from rasa.engine.storage.storage import ModelStorage
+from rasa.engine.training.components import (
+    CachedComponent,
+    FingerprintComponent,
+    FingerprintStatus,
+)
+from rasa.engine.training.hooks import TrainingHook
+from rasa.shared.core.domain import Domain
+
+
+logger = logging.getLogger(__name__)
+
+
+class GraphTrainer:
+    """Trains a model using a graph schema."""
+
+    def __init__(
+        self,
+        model_storage: ModelStorage,
+        cache: TrainingCache,
+        graph_runner_class: Type[GraphRunner],
+    ) -> None:
+        """Initializes a `GraphTrainer`.
+
+        Args:
+            model_storage: Storage which graph components can use to persist and load.
+                Also used for packaging the trained model.
+            cache: Cache used to store fingerprints and outputs.
+            graph_runner_class: The class to instantiate the runner from.
+        """
+        self._model_storage = model_storage
+        self._cache = cache
+        self._graph_runner_class = graph_runner_class
+
+    def train(
+        self,
+        train_schema: GraphSchema,
+        predict_schema: GraphSchema,
+        domain_path: Path,
+        output_filename: Path,
+    ) -> GraphRunner:
+        """Trains and packages a model and returns the prediction graph runner.
+
+        Args:
+            train_schema: The train graph schema.
+            predict_schema: The predict graph schema.
+            domain_path: The path to the domain file.
+            output_filename: The location to save the packaged model.
+
+        Returns:
+            A graph runner loaded with the predict schema.
+
+        """
+        logger.debug("Starting training.")
+
+        pruned_training_schema = self._fingerprint_and_prune(train_schema)
+
+        hooks = [TrainingHook(cache=self._cache, model_storage=self._model_storage)]
+
+        graph_runner = self._graph_runner_class.create(
+            graph_schema=pruned_training_schema,
+            model_storage=self._model_storage,
+            execution_context=ExecutionContext(graph_schema=pruned_training_schema),
+            hooks=hooks,
+        )
+
+        logger.debug("Running the pruned train graph with real node execution.")
+
+        graph_runner.run()
+
+        domain = Domain.from_path(domain_path)
+        model_metadata = self._model_storage.create_model_package(
```

Comment:
> Is the "I want to change these parameters for this prediction run" feature already included somewhere?

Not here (at the moment). That should be done in the graph loader.
This one is really just in case just want to use a model directly after training 

>  but there's no good reason to separate those 🤔

Why not? Loading really just loads it while graph trainer does so much more stuff (fingerprinting, pruning, packaging)🤔

Reply:
> Why not?
I just didn't want to cause too much unnecessary rework with my comments :D Conceptually, I'd think it would be great to separate those two things

Reply:
Ah, sorry, I've misunderstood you. I thought you'd prefer to merge them. So you're saying that the graph trainer should not return a loaded model at all to not duplicate work from the "graph loader"?

Reply:
yep - but I guess there's no need to e.g. have two predict_graph definitions (in that case separating those two things would be nice)

Reply:
I think it's a convenience functionality for now that the trainer returns the instantiated model. We should check in the end if we are actually using this or if we can just return `None`